Feature/domain exception setup

kdy2224 opened the pull request:
-exception 세팅 추가
-domain 구현, jpa 매핑

kdy2224 committed:
test commit

kdy2224 committed:
exception 세팅 추가

kdy2224 committed:
도메인 구현 및 jpa 매핑처리
-GameRoomSetting -> GameSetting(클래스다이어그램과 일치)으로 변경
-domain 에 BaseEntity 추상 클래스 추가
-many to many 는 별도 매핑 테이블로 뺌

kdy2224 committed:
chore: 누락된 코드 추가
-몇몇 엔티티 클래스에 BaseEntity 상속 코드 추가

kdy2224 committed:
엔티티 클래스에 메서드 추가

kdy2224 commented:
각 도메인 클래스에서 어떻게 구현해야 할지 모르겠는 메서드는 아직 구현이 안돼있습니다.(주석)

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/ChatMessage.java:
```diff
@@ -0,0 +1,62 @@
+	private String content;
+
+	@CreatedDate
+	private LocalDateTime timestamp;
```
BaseEntity의 createAt과 timestamp의 역할이 겹치지 않을까요?

kdy2224 replied:
아 겹치네요.! 저는 메세지를 보내는 시간을 저장하는 필드로 생각해서 객체가 만들어지는 시점과 별도로 생각했어요. 그럼 @CreatedDate 를 빼야겠네요.!

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/GameCardPool.java:
```diff
@@ -0,0 +1,39 @@
+		if (isEmpty()) {
+			return null; //예외처리
+		}
+		return cards.remove(cards.size() - 1);
```
저는 List 형태가 아닌 Queue 형태라고 생각했습니다. 왜냐하면 카드 풀에서 카드를 한장씩 빼는 그림이라 생각했거든요.
도연님은 어떻게 생각하시나요?

kdy2224 replied:
이부분은 생각을 잘못해서 말씀하신대로 수정했습니다.!

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/GameLog.java:
```diff
@@ -0,0 +1,62 @@
+		sb.append("Game ID: ").append(gameId).append("\n");
+		sb.append("Start Date: ").append(startDate).append("\n");
+		sb.append("End Date: ").append(endDate).append("\n");
+		sb.append("Users: ").append(users).append("\n");
```
users라는 배열을 StringBuilder에 붙이면 스트링형태로 자동으로 변환해주나요?
변환이 된다면 어떤 형태로 변환되나요?

kdy2224 replied:
User 객체의 toString() 을 , 로 연결한 형태요..! 따로 toString() 을 예쁘게 만들어둬도 좋을거 같습니다.

jiho5993 replied:
저는 raw log를 만들때 저희가 이 로그를 사용해서 어딘가에 가공할 수 있는 형태로 만드는게 좋다는 생각이 들어요.
예를 들어, json 형태로 아래처럼 만들었다고 했을때,
```json
{
    "gameId": 1,
    "startDate": "2025-05...",
    "endDate": "2025-05...",
    "users": [  ],
    "roundLogs": [  ]
}
```
이 데이터를 사용해서 프론트 화면에 동적으로 보여질 수 있도록 가공한다던지, 아니면 저희만의 로그 시스템을 구축할 때 가공한다던지 등등..
최종적으론 저 json 데이터를 String화 해서 리턴하는 방식을 생각했었어요.

그래서 제 생각에는 지금 방식은 아래처럼 나올 것 같은데, 이 방식은 재활용 불가능한 형태의 로그라고 생각하거든요.
```
Game ID: 1
Start Date: "2025-05..."
End Date: "2025-05..."
Users: ...
Round Logs: ...
```

혹시 지금의 방식으로는 저 리턴되는 로그를 어디에 어떻게 사용할 수 있는지? 이런게 궁금합니다.
도연님은 어떻게 생각하시나요? 편하게 남겨주세요!

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/GameRoom.java:
```diff
@@ -1,31 +1,62 @@
+
+    private Integer round = 1;
+
+    @OneToOne(mappedBy = "gameRoom", cascade = CascadeType.ALL)
```
cascade ALL로 설정하면 GameRoom을 삭제했을때 GameSetting도 자동으로 같이 삭제되지 않나요?
그러면 연관되어 있는 엔티티들 모두 삭제될 것 같은데, 필요한 것만 삭제하는 것이 좋을 것 같다는 생각이 듭니다.

kdy2224 replied:
GameRoom이 삭제되면 해당 GameSetting 도 필요 없을거라 생각했긴 한데 일단 CascadeType.ALL 은 빼둘게요..!

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/RoleAssignment.java:
```diff
@@ -0,0 +1,35 @@
+@Entity
+@AllArgsConstructor
+@NoArgsConstructor
+public class RoleAssignment extends BaseEntity {
```
RoleAssignment말고 GameCardAssignment처럼 Game이라는 prefix를 붙여서 네이밍을 통일시키는게 좋을 것 같네요.

kdy2224 replied:
반영했습니다.!

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/Timer.java:
```diff
@@ -0,0 +1,32 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Embeddable
+public class Timer extends BaseEntity {
```
Timer도 GameTimer로 바꾸는게 좋을 것 같아요.

kdy2224 replied:
반영했습니다.!

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/UserCardDeck.java:
```diff
@@ -0,0 +1,32 @@
+@Entity
+@AllArgsConstructor
+@NoArgsConstructor
+public class UserCardDeck extends BaseEntity {
```
나중에 카드를 내는 로직도 하나 만들어야겠네요

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/enums/UserLoginType.java:
```diff
@@ -0,0 +1,5 @@
+package com.goldstone.saboteur_backend.domain.enums;
+
+public enum UserLoginType {
+	GOOGLE, APPLE, FACEBOOK, EMAIL, ANONYMOUS
```
Email이랑 Anonymous는 어떤 차이가 있는건가요?
Anonymous는 비로그인 유저들을 위한 기능인가요?

kdy2224 replied:
이부분 무슨 의미인줄 몰라서 일단 이렇게 뒀습니다.! UserLoginType 이 OAuth에서 제공자(외부 서비스)를 의미하는게 맞을까요?

jiho5993 replied:
넵넵 카카오 유저, 구글 유저, 아니면 소셜 로그인 없이 직접 회원가입한 유저 등등.. 나눌 생각이었습니다.

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/exception/code/error/CommonErrorCode.java:
```diff
@@ -0,0 +1,19 @@
+
+	INVALID_REQUEST(HttpStatus.BAD_REQUEST, "C001", "잘못된 요청입니다."),
+	INVALID_INPUT(HttpStatus.BAD_REQUEST, "C002", "입력값이 올바르지 않습니다."),
+	INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "S001", "서버 오류입니다.");
```
C랑 S는 어떤 차이가 있는건가요?

kdy2224 replied:
CommonError 라서 c 로 두려고 한건데 s 가 들어간 건 잘못 들어갔네요.!

jiho5993 commented:
`./gradlew spotlessApply` 명령어로 lint 처리부탁드려요!

kdy2224 committed:
리뷰 반영

kdy2224 commented:
lint 처리해서 이번 커밋에 바뀐게 많습니다. 댓글 달아주셨던 파일 위주로 봐주세요.!

kdy2224 committed:
chore: 오타(오류) 수정

kdy2224 committed:
application.example.yml 수정(jpa부분)

kdy2224 committed:
chore: application.example.yml 수정

kdy2224 committed:
GameLog.java-게임로그 반환 메서드 수정

jiho5993 commented:
GameRoundLog, GameRoundUserLog도 수정해주세요!

kdy2224 committed:
GameRoundUserLog.java, GameRoundLog.java 의 로그 반환 메서드 수정

jiho5993 commented on src/main/java/com/goldstone/saboteur_backend/domain/GameLog.java:
```diff
@@ -0,0 +1,76 @@
+
+    private Long gameId;
+
+    @OneToMany private List<User> users;
```
jpa 설정해놓으셔서 직접 실행해보고 테이블 잘 생성되나 확인해봤는데,
밑의 GameRoundLog는 GameLog와 외래키가 잘 설정되어 있어서 참조가 가능하지만,
User는 설정이 안되어 있어서 외래키 생성은 물론 참조도 안되네요.

jpa에 대해 잘 알지 못해서  따로 확인해봤더니 ManyToOne 어노테이션을 유저테이블에 추가해야할 것 같은데, 의도된 부분일까요?

kdy2224 committed:
엔티티 클래스 수정
-GameRoundLog.java, GameRoundUserLog.java 의 로그 반환 메서드 수정
-OneToMany 단방향 관계애서 외래키가 생성 안되는 부분 수정

kdy2224 committed:
엔티티 클래스 수정
-모든 1:N 관계의 단방향 매핑을 양방향으로 수정

jooho172200 commented on src/main/java/com/goldstone/saboteur_backend/domain/UserRank.java:
```diff
@@ -27,4 +28,8 @@ public class UserRank extends BaseEntity {
     private Integer rank;
```
실행 후 콘솔 로그를 보니 rank가 MySQL 예약어라 테이블 생성에 오류를 일으키는 거 같아 @Column(name = "`rank`")를 붙여서 해결했습니다.

jiho5993 replied:
rank말고 다른 이름은 없을까요? 예약어랑 겹치는 거라면 최대한 피하는게 좋을 것 같아서요
